php-frontend/src/main/java/org/sonar/php/filters/SuppressWarningFilter.java

```diff
@@ -0,0 +1,124 @@
+import org.sonar.plugins.php.api.tree.lexical.SyntaxTrivia;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
+
+public class SuppressWarningFilter extends PHPVisitorCheck implements FilterPhpIssue {
+
+  private final Map<String, Map<Integer, Set<String>>> suppressedRulesPerLinePerFile = new HashMap<>();
+  private String fileUri;
+
+  private static final String ARGUMENT_FORMAT = "\"[a-zA-Z0-9:]++\"";
+  private static final String ARGUMENTS_FORMAT = ARGUMENT_FORMAT + "(?:\\s*+,\\s*+"+ARGUMENT_FORMAT+")*+";
+  private static final Pattern SUPPRESS_WARNING_COMMENT_PATTERN = Pattern.compile("@SuppressWarnings\\s*+\\(\\s*+(?<arguments>"+ARGUMENTS_FORMAT+")\\s*+\\)");
+
+  public void reset() {
+    suppressedRulesPerLinePerFile.clear();
+  }
+
+  public void scanCompilationUnit(String fileUri, CompilationUnitTree tree) {
+    this.fileUri = fileUri;
+    super.visitCompilationUnit(tree);
+  }
+
+  @Override
+  public boolean accept(String fileName, String ruleName, int line) {
+    return !suppressedRulesPerLinePerFile.getOrDefault(fileName, Collections.emptyMap())
+      .getOrDefault(line, Collections.emptySet()).contains(ruleName);
+  }
+
+  @Override
+  public void visitAttribute(AttributeTree tree) {
+    Set<String> rulesSuppressed = tree.arguments().stream()
+      .map(CallArgumentTree::value)
+      // consider only string literal
+      .filter(expr -> expr.is(Tree.Kind.REGULAR_STRING_LITERAL)).map(LiteralTree.class::cast)
+      .map(LiteralTree::value)
+      .map(SuppressWarningFilter::stripDoubleQuotes)
+      .collect(Collectors.toSet());
+    suppressedRulesPerLinePerFile.computeIfAbsent(fileUri, key -> new HashMap<>()).put(getAttributeLine(tree), rulesSuppressed);
+    super.visitAttribute(tree);
+  }
+
+  public int getAttributeLine(AttributeTree tree) {
+    // TODO: SONARPHP-1368 replace this logic "ignore next line issue" by more advanced "the scope of parent group attribute"
+    AttributeGroupTree parent = (AttributeGroupTree) tree.getParent();
+    return parent.endToken().endLine()+1;
+  }
+
+  @Override
+  public void visitToken(SyntaxToken token) {
+    for (SyntaxTrivia trivia : token.trivias()) {
+      String comment = getContents(trivia.text());
+      processSuppressedWarningsInComment(token, comment);
+    }
+    super.visitToken(token);
+  }
+
+  private void processSuppressedWarningsInComment(SyntaxToken token, String comment) {
+    Matcher matcher = SUPPRESS_WARNING_COMMENT_PATTERN.matcher(comment);
+    while (matcher.find()) {
+      String arguments = matcher.group("arguments");
+      Arrays.stream(arguments.split(","))
+        .map(str -> stripDoubleQuotes(str.trim()))
+        .forEach(ruleName -> {
+          for (int line = token.line(); line <= token.endLine(); line++) {
+            suppressedRulesPerLinePerFile.computeIfAbsent(fileUri, key -> new HashMap<>())
+              .computeIfAbsent(line, key -> new HashSet<>()).add(ruleName);
```

Comment:
In general I would avoid chaining these calls, as they are quite hard to read.
I would suggest using intermediate variables after the `computeIfAbsent` calls.

Reply:
All in all there's a whole lot of complexity hidden in this lambda, which would benefit from breaking it down.

Reply:
I agree with this, this is also strongly related to the fact that it is using a Map of Map of...
I was planning to change this to a better structure in the second part of the ticket, but I will try to come up with something better now.